### PR TITLE
Forking MVP

### LINKED ISSUE #344: Starknet state forking - MVP
It should be possible to load starknet state locally and run tests on it instead of the fresh local instance

### PR DESCRIPTION
**Pull request by war-in**
## Introduced changes

- contracts deployed on real networks can be accessed in the tests now

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/forking/state.rs
Comment (Arcticae) [inline crates/cheatnet/src/forking/state.rs:16]: I think we should store the values in dictionaries (a simple cache without storing anything on disk)

Comment (Arcticae) [reply crates/cheatnet/src/forking/state.rs:16]: I think we're reading right now from the dictionary, but we're not writing to it on response before returning the value

Comment (Arcticae) [reply crates/cheatnet/src/forking/state.rs:16]: Or does it cache automatically after returning? I might be wrong here

Comment (war-in) [reply crates/cheatnet/src/forking/state.rs:16]: Yup, it is cached automatically in the CachedState which calls the `CustomStateReader`

File: crates/cheatnet/tests/common/state.rs
Comment (war-in) [reply crates/cheatnet/tests/common/state.rs:17]: So should I use `.env` file or maybe some other approach?

File: crates/cheatnet/src/state.rs
Comment (war-in) [reply crates/cheatnet/src/state.rs:122]: They are in the correct order now

File: crates/cheatnet/tests/cheatcodes/mod.rs
Comment (Radinyn) [inline crates/cheatnet/tests/cheatcodes/mod.rs:9]: 1. Forking should be moved to a different testing module or at least integrated with the starknet module as it is not a cheatcode from the Cheatnet's perspective and we should be able to run it's test separately.

**Review by Radinyn — APPROVED** [review]

File: crates/cheatnet/src/lib.rs
Comment (Arcticae) [inline crates/cheatnet/src/lib.rs:33]: I still think that `StateReader` should be provided from outside rather than passing optional RPC conf

Comment (Arcticae) [reply crates/cheatnet/src/lib.rs:33]: If you dislike that approach i suggest as well an alternative (which i kind of dislike as well, but not as much as this one):

Add another constructor method which instantiates the whole thing, but name it `from_rpc` or `new_forked` and then we can differentiate on how we instantiate those.

That said - I still prefer injecting such deps since it's cleaner and this code will actually compose more nicely with changes that will inevitably be made to it (multifork)

File: crates/cheatnet/src/state.rs
Comment (Arcticae) [inline crates/cheatnet/src/state.rs:86]: This is still moved

File: crates/cheatnet/tests/common/state.rs
Comment (Arcticae) [inline crates/cheatnet/tests/common/state.rs:16]: It doesn't seem to belong in this method

File: .gitignore
Comment (Arcticae) [inline .gitignore:6]: We should add info about this to developers guide

**Review by Arcticae — CHANGES_REQUESTED** [review]
Please refrain from merging before Maks has taken a look

File: crates/cheatnet/tests/starknet/forking.rs
Comment (Arcticae) [inline crates/cheatnet/tests/starknet/forking.rs:1]: We are lacking tests which check the state differences between different block heights

**Comment by Arcticae** [comment]
Also - let's not close this issue with this PR, we will need to follow-up on forge's side, etc.

**Review by piotmag769 — CHANGES_REQUESTED** [review]
What is the planned approach for testing it? Will we host our own RPC node?

**Comment by Radinyn** [comment]
> What is the planned approach for testing it? Will we host our own RPC node?

We already do for testing SDK's

File: crates/cheatnet/src/forking/worker.rs
Comment (MaksymilianDemitraszek) [inline crates/cheatnet/src/forking/worker.rs:31]: It would be probably better to keep one runtime, as afaik creating new runtime spawns a thread each time.

**Review by MaksymilianDemitraszek — DISMISSED** [review]
LGTM, amazing job with scoping the MVP, we probably can progress with forge implementation rn and keep working on the optimization on the cheatnet side. 
I will finish review Monday morning as I am a bit tired rn 😄 

Aside, @piotmag769 @Radinyn and maybe @THenry14 (i think you considered it for cast) we have to be careful with testing using hosted node as it will be very problematic for contributors to replicate 😢 . We probably should consider other options first.

File: crates/cheatnet/src/forking/state.rs
Comment (piotmag769) [reply crates/cheatnet/src/forking/state.rs:16]: Unresolving, since this looks like a major concern to me
cc @MaksymilianDemitraszek

File: crates/cheatnet/src/forking/worker.rs
Comment (war-in) [reply crates/cheatnet/src/forking/worker.rs:23]: I'm not sure. We keep runtime inside it so `Worker` is ok now

File: crates/cheatnet/src/forking/state.rs
Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/forking/state.rs:16]: > low prob that someone will want to use contract deployed at address e.g. 0x101

@piotmag769 what do you mean by that?

File: crates/cheatnet/src/forking/worker.rs
Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/forking/worker.rs:23]: I mean it is not a strong opinion and it is fine name but imo it is a little weird

File: crates/cheatnet/src/state.rs
Comment (MaksymilianDemitraszek) [inline crates/cheatnet/src/state.rs:21]: StateReaderProxy/Router maybe would be a better name?

Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/state.rs:21]: What is the performance overhead of keeping it on the heap? I assume nothing significant but I wanted to raise it so you can think about it.

**Review by MaksymilianDemitraszek — COMMENTED** [review]
Do we have tests wich verify that all possible state modifications work as intended?

File: .github/workflows/ci.yml
Comment (cptartur) [reply .github/workflows/ci.yml:77]: Also will this work for external PRs at all?

Comment (THenry14) [reply .github/workflows/ci.yml:77]: > Also will this work for external PRs at all?
Depends on what you count as an external PR. If someone forks and runs CI on the fork, the secret is not available.
When someone gets an approval to run CI on our repo - this is where things start to get tricky. We must make sure the output is censored (I'm pretty positive it is) and that no shenanigans in the form of redirecting the secret to file and downloading it works. Pretty solid PRs before allowing the runs should suffice, but we must make sure that those are needed after every new commit (otherwise someone that got such approval could try to do the file redirect thing).

Comment (cptartur) [reply .github/workflows/ci.yml:77]: I think the current flow is that only approval for the first run is needed (actually its for the first PR, but it can be changed) and then the subsequent runs just work. So this is not ideal.

Though [the docs say](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) "Although workflows from forks do not have access to sensitive data such as secrets, they can be an annoyance for maintainers if they are modified for abusive purposes." so it seems this will not work with PRs from forks at all. But it should be verified probably.

File: crates/cheatnet/tests/starknet/forking.rs
Comment (piotmag769) [inline crates/cheatnet/tests/starknet/forking.rs:27]: What address is it? Does calling `increase_balance` not mutate the state? Won't it cause the test to fail when a new block is minted?

File: crates/cheatnet/tests/starknet/forking.rs
Comment (piotmag769) [inline crates/cheatnet/tests/starknet/forking.rs:88]: Wouldn't it be better to have a constructor method that takes block number as arg?

File: crates/cheatnet/tests/starknet/forking.rs
Comment (piotmag769) [inline crates/cheatnet/tests/starknet/forking.rs:79]: Do I understand correctly that you manually checked the block number of the block that contract with the address was included in?

File: crates/cheatnet/tests/starknet/forking.rs
Comment (war-in) [reply crates/cheatnet/tests/starknet/forking.rs:27]: This is the `HelloStarknet` contract address deployed by me for this specific test suite.
`increase_balance` mutates the state but only locally so it will not affect future runs

File: crates/cheatnet/tests/starknet/forking.rs
Comment (war-in) [reply crates/cheatnet/tests/starknet/forking.rs:88]: Do you mean the `ForkStateReader` constructor?

File: crates/cheatnet/tests/starknet/forking.rs
Comment (war-in) [reply crates/cheatnet/tests/starknet/forking.rs:79]: Yup

File: crates/cheatnet/tests/starknet/forking.rs
Comment (piotmag769) [reply crates/cheatnet/tests/starknet/forking.rs:27]: Oh true, forgot we only mutate it locally, thanks!

File: .github/workflows/ci.yml
Comment (war-in) [reply .github/workflows/ci.yml:77]: Conclusion:
- every new commit requires approval
- amended commits too
- force-pushed branch also

We can safely keep our node address in the secrets

File: crates/cheatnet/src/state.rs
Comment (war-in) [reply crates/cheatnet/src/state.rs]: I don't think it will be much more heavier than the one without forking

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Pull request merged**
